frontend/src/metabase/ui/components/data-display/TreeTable/hooks/useColumnSizing.tsx

```diff
@@ -0,0 +1,413 @@
+        widths[colId] = minConstraint;
+        remainingSpace -= minConstraint;
+        stretchingColumnIds.delete(colId);
+        anyConstrained = true;
+        break;
+      }
+
+      if (column.maxWidth != null && baseWidth > column.maxWidth) {
+        widths[colId] = column.maxWidth;
+        remainingSpace -= column.maxWidth;
+        stretchingColumnIds.delete(colId);
+        anyConstrained = true;
+        break;
+      }
+    }
+
+    if (!anyConstrained) {
+      for (const colId of stretchingColumnIds) {
+        widths[colId] = baseWidth;
+      }
+      break;
+    }
+  }
+
+  return widths;
+}
+
+interface MeasureContentProps<TData extends TreeNodeData> {
+  columns: TreeTableColumnDef<TData>[];
+  rows: Row<TData>[];
+  table: Table<TData>;
+  hasExpandableNodes: boolean;
+  indentWidth: number;
+  onMeasured: (widths: Record<string, number>, maxDepth: number) => void;
+}
+
+function MeasureContent<TData extends TreeNodeData>({
+  columns,
+  rows,
+  table,
+  hasExpandableNodes,
+  indentWidth,
+  onMeasured,
+}: MeasureContentProps<TData>) {
+  const handleRef = useCallback(
+    (div: HTMLDivElement | null) => {
+      if (!div) {
+        return;
+      }
+
+      const widths: Record<string, number> = {};
+      let maxDepth = 0;
+
+      for (const row of rows) {
+        maxDepth = Math.max(maxDepth, row.depth);
+      }
+
+      columns.forEach((column, colIndex) => {
+        const columnEl = div.querySelector(
+          `[data-measure-column="${column.id}"]`,
+        );
+        if (!columnEl) {
+          return;
+        }
+
+        const cellEls = columnEl.querySelectorAll("[data-measure-cell]");
+        let maxWidth = 0;
+
+        cellEls.forEach((cellEl) => {
+          const depth = parseInt(cellEl.getAttribute("data-depth") ?? "0", 10);
+          let cellWidth = (cellEl as HTMLElement).offsetWidth;
+
+          cellWidth += CELL_HORIZONTAL_PADDING;
+
+          if (colIndex === 0) {
+            cellWidth += TREE_CELL_BASE_PADDING;
+            if (hasExpandableNodes) {
+              cellWidth += EXPAND_BUTTON_WIDTH;
+            }
+            cellWidth += depth * indentWidth;
+          }
+
+          maxWidth = Math.max(maxWidth, cellWidth);
+        });
+
+        const headerEl = columnEl.querySelector("[data-measure-header]");
+        if (headerEl) {
+          const headerWidth =
+            (headerEl as HTMLElement).offsetWidth + CELL_HORIZONTAL_PADDING;
+          maxWidth = Math.max(maxWidth, headerWidth);
+        }
+
+        widths[column.id] = maxWidth;
+      });
+
+      onMeasured(widths, maxDepth);
+    },
+    [columns, rows, hasExpandableNodes, indentWidth, onMeasured],
+  );
+
+  const columnsToMeasure = columns.filter(needsMeasurement);
+
+  return (
+    <div ref={handleRef} style={{ display: "flex" }}>
+      {columnsToMeasure.map((column) => {
+        const colIndex = columns.findIndex((c) => c.id === column.id);
+        const colAccessorFn =
```

Comment:
Rather than compute an accessFn based off the column defintion, could we just use `row[i].getValue()`?